### DISCUSSION TITLE
Online: Scoping Out of Catchment Area in FHIR

### DISCUSSION THREAD

**Discussion by f-odhiambo** [opensrp/fhircore#2627] [Ideas]
# Managing Patients Out of Catchment Area  - FHIR-Based Approach

**Out of Catchment Types**
There are two main types of Out of Catchment movement i.e. 
1. **Out of Catchment** - Patient belongs to location A,  receives services in Location B , creates/updates some records based on where they receive the healthcare service 
2. **Permanent**  - Patient belongs to location A, moves to Location B to receive services  - Transfer all records, change cares team 

Additionally, there are also other services launched via the Overflow Menu in the Patient profile, such as the listed below, which will be covered later but is out of the scope currently.

1. **Out of Area** - No record exists - You come to use it to seek services from an out-of-catchment facility. The patient knows their system identifier. The clinician saves a vaccine using the identifier. The system is able to reconcile the vaccine record at the patient's home facility.
2. **Mobile/Outreach**  - We come to you instead of you coming to us 

# Sample Workflows from OpenSPR I

**Scenarios:**

* There are two clinics, Clinic A and B.
* Clinic A is Susan’s Home Health Facility as defined in the Child Registration Form
* Susan goes to Clinic A and receives the birth and 6-week immunizations.
* Susan’s Record is updated with her latest immunization record.
* Susan attends Clinic B for her 10-week immunizations because it’s more convenient

Sunny Path Workflow (Matching Identifier) :

* Susan goes to Clinic B, and her record is imported
* The clinician sees that she is due for her 10-week immunizations
* The clinician orders those immunizations and updates Susan’s record
* The immunizations that were given today are synchronized from Clinic B to the OpenSRP HAPI FHIR server
* The OpenSRP HAPI FHIR server sees that Susan’s home health facility is Clinic A and adds those immunizations to the queue of records that need to be synchronized with Clinic A
* Clinic A syncs their tablet, and Susan’s updated record is downloaded

Rainy Path Workflow (No Matching Identifier) :

* Susan’s caregiver states that she doesn’t have her HealthCard Card with her
* The clinician tries to search their tablet for Susan’s record but doesn’t find it
* The clinician tries to perform a global search using the Advanced Search
    * If online, the clinician can search for Susan’s record:
        * The clinician asks Susan if they want to make this facility their home health facility.
            * If yes, the clinician imports the record and proceeds
            * If no, the clinician has to complete the out-of-catchment form as if they were not online
        * The clinician sees that she is due for her 10-week immunizations
        * The clinician orders those immunizations and reissues the card
        * The OpenSRP HAPI FHIR server sees that Susan’s home health facility is now Clinic B and removes her record from future syncing at Clinic A
    * If offline(Assumes we have already retrieved the user's record above and saved to device)
        * Complete the immunizations that were given today(via the user's profile).
        * When the connection resumes, this data syncs from Clinic B to the OpenSRP HAPI FHIR server
        * The child’s record is matched, and the immunization is added to the Child’s health record at Clinic A
  
# Proposed  Solution 

**Assumptions**
1. All data will be generated from the OpenSPR FHIR Core app; thus, all the patients will have the same format and also a similar structure of CarePlan & Tasks if $Lite is invoked at registration 
2. All generated data will be either stored locally or residing from a HAPI FHIR Server, where it will also be synced to an app on the initial setup 
3. A sync strategy will be in place for any of the apps pulling data from the HAPI FHIR Server 

**Steps to follow for the walkthrough:** 

**1. Perform a  Remote search** 
1. Retrieve Patient data by searching by  Patient Identifier 
2. Find Patient resources by ID. These could be multiple, but we need the demographic details returned so the user can select them in the case of a fuzzy search.
3. If a match is found, then do Patient/ID/$everthing to get all data 

**Pre-requisite** 
**BLOCKER**  - Perform Remote Search as an entry point  - https://github.com/opensrp/fhircore/discussions/2163 

**2. Manage Patient Data** 
Once the data has been successfully synced, we can begin updating/creating new records. Some resources to be used to track this include 
1. Encounter - Mark that an out-of-catchment transfer was made 
2. Flag - Tag the patient to which the move was made and whether it was a permanent or a temporary move, and also capture the duration for when one was receiving care out of the area
3. Update/Generate any new data  - Update Task status, Create an Encounter
4. Store all newly generated data locally till remote sync is next performed 

**Open Questions** 
1. How does this affect the stock reporting (e.g., vaccines)? 
2. How do we handle the RelatedPerson (Relatives/Guardian) or the Family(Group) data when a client moves?
3. In the case of a temporary move, do we need to delete after the prescribed expiry criteria? If they remain, what happens? 
4. Patient Record Reconciliation  
        - We need to perform deduplication and merge of records on HAPI once records are synced successfully 
        - We also need to scope the rules for reconciling/conflict resolution on the server side. What becomes the source of truth LastUpdate? 

**Sample Patient Resource** 
```json
{
    "resourceType": "Patient",
    "id": "61f440c5-fcc6-480f-bfe0-e49bfab71f08",
    "meta": {
        "versionId": "1",
        "lastUpdated": "2023-06-14T11:35:17.802+00:00",
        "source": "#4a2e50b69708cf29",
        "tag": [
            {
                "system": "https://smartregister.org/app-version",
                "code": "0.2.2-zeir",
                "display": "Application Version"
            },
            {
                "system": "https://smartregister.org/location-tag-id",
                "code": "Not defined",
                "display": "Practitioner Location"
            },
            {
                "system": "https://smartregister.org/organisation-tag-id",
                "code": "1a8ae4f2-c94b-4366-ac31-f55ccf3168a2",
                "display": "Practitioner Organization"
            },
            {
                "system": "https://smartregister.org/care-team-tag-id",
                "code": "2309fc1e-9fbc-46f3-8a70-8dbd9d2448ed",
                "display": "Practitioner CareTeam"
            },
            {
                "system": "https://smartregister.org/practitioner-tag-id",
                "code": "6e744577-cf1c-4876-a122-2d0e9c5e0f23",
                "display": "Practitioner"
            }
        ]
    },
    "identifier": [
        {
            "use": "usual",
            "value": "2286466389"
        }
    ],
    "active": true,
    "name": [
        {
            "use": "official",
            "text": "A",
            "family": "Jane",
            "given": [
                "Doe"
            ]
        }
    ],
    "gender": "female",
    "birthDate": "2023-06-02",
    "generalPractitioner": [
        {
            "reference": "Practitioner/6e744577-cf1c-4876-a122-2d0e9c5e0f23"
        }
    ],
    "managingOrganization": {
        "reference": "Organization/1a8ae4f2-c94b-4366-ac31-f55ccf3168a2"
    }
}
````

**Sample Encounter**

````json
{
   "resourceType": "Encounter",
   "id": "020cbc35-d496-4f5d-8918-417330cd9b69",
   "meta": {
      "versionId": "1",
      "lastUpdated": "2023-08-01T07:28:31.086+00:00",
      "source": "#eb8506f6340fecae",
      "tag": [
         {
            "system": "https://smartregister.org/organisation-tag-id",
            "code": "1a8ae4f2-c94b-4366-ac31-f55ccf3168a2",
            "display": "Practitioner Organization"
         },
         {
            "system": "https://smartregister.org/care-team-tag-id",
            "code": "2309fc1e-9fbc-46f3-8a70-8dbd9d2448ed",
            "display": "Practitioner CareTeam"
         },
         {
            "system": "https://smartregister.org/app-version",
            "code": "0.2.4-zeir",
            "display": "Application Version"
         },
         {
            "system": "https://smartregister.org/practitioner-tag-id",
            "code": "6e744577-cf1c-4876-a122-2d0e9c5e0f23",
            "display": "Practitioner"
         },
         {
            "system": "https://smartregister.org/location-tag-id",
            "code": "Not defined",
            "display": "Practitioner Location"
         }
      ]
   },
 "contained": [
    {
      "resourceType": "Loca",
      "id": "home",
      "description": "New Location",
      "mode": "kind"
    }
  ],
   "status": "finished",
   "class": {
      "system": "http://terminology.hl7.org/CodeSystem/v3-ActCode",
      "code": "HH"
   },
   "type": [
      {
         "coding": [
              {
               "system": "http://smartregsiter.org/",
               "code": "000112",
               "display":"Out of Catchment Area Move"
            }
         ],
         "text": "Out of Catchment Area Move"
      }
   ],
   "priority": {
      "coding": [
         {
            "system": "http://terminology.hl7.org/ValueSet/v3-ActPriority",
            "code": "EL",
            "display": "elective"
         }
      ],
      "text": "elective"
   },
   "subject": {
      "reference": "Patient/61f440c5-fcc6-480f-bfe0-e49bfab71f08"
   },
   "participant": [
      {
         "individual": {
            "reference": "Practitioner/6e744577-cf1c-4876-a122-2d0e9c5e0f23"
         }
      }
   ],
   "period": {
      "start": "2023-07-05T12:17:56+03:00",
      "end": "2023-07-05T12:17:56+03:00"
   },
   "reasonCode": [
      {
         "coding": [
            {
               "system": "http://smartregsiter.org/",
               "code": "000112",
               "display":"Out of Catchment Area Move"
            }
         ],
         "text": "Out of Catchment Area Move"
      }
   ],
   "serviceProvider": {
      "reference": "Organization/1a8ae4f2-c94b-4366-ac31-f55ccf3168a2"
   },
 "location": [
    {
      "location": {
        "reference": "#loca",
        "display": "New Location"
      },
      "status": "completed",
      "period": {
        "start": "2015-01-17T16:00:00+10:00",
        "end": "2015-01-17T16:30:00+10:00"
      }
    }
  ]
}
````

**Sample Flags** 

**Permanent Move** 
````json 
{
  "resourceType": "Flag",
  "id": "example",
  "status": "active",
  "code": {
    "coding": [
      {
        "system": "http://www.snomed.ct",
        "code": "0001122",
        "display": "OutofCatchment"
      }
    ],
    "text": "Permanent Move"
  },
  "subject": {
    "reference": "Patient/61f440c5-fcc6-480f-bfe0-e49bfab71f08"
  },
  "period": {
      "start": "2023-07-05T12:17:56+03:00",
      "end": "2023-07-05T12:17:56+03:00"
  },
  "author": {
    "reference": "Practitioner/6e744577-cf1c-4876-a122-2d0e9c5e0f23"
  },
  "encounter": {
    "reference": "Encounter/example"
  }
}
````

**Temporary Move** 
````json 
{
  "resourceType": "Flag",
  "id": "example",
  "status": "active",
  "code": {
    "coding": [
      {
        "system": "http://www.snomed.ct",
        "code": "0001123",
        "display": "OutofCatchment"
      }
    ],
    "text": "Temporary Move"
  },
  "subject": {
    "reference": "Patient/61f440c5-fcc6-480f-bfe0-e49bfab71f08"
  },
  "period": {
       "start": "2023-07-05T12:17:56+03:00",
      "end": "2023-07-05T12:17:56+03:00"
  },
  "author": {
    "reference": "Practitioner/6e744577-cf1c-4876-a122-2d0e9c5e0f23"
  },
  "encounter": {
    "reference": "Encounter/example"
  }
}
````

**Out-of-Area Service** 
````json 
{
  "resourceType": "Flag",
  "id": "example",
  "status": "active",
  "code": {
    "coding": [
      {
        "system": "http://www.snomed.ct",
        "code": "0001124",
        "display": "OutofCatchment"
      }
    ],
    "text": "Out-of-Area Service"
  },
  "subject": {
    "reference": "Patient/61f440c5-fcc6-480f-bfe0-e49bfab71f08"
  },
  "period": {
      "start": "2023-07-05T12:17:56+03:00",
      "end": "2023-07-05T12:17:56+03:00"
  },
  "author": {
    "reference": "Practitioner/6e744577-cf1c-4876-a122-2d0e9c5e0f23"
  },
  "encounter": {
    "reference": "Encounter/example"
  }
}
````

**Outreach Service** 
````json 
{
  "resourceType": "Flag",
  "id": "example",
  "status": "active",
  "code": {
    "coding": [
      {
        "system": "http://www.snomed.ct",
        "code": "0001125",
        "display": "OutofCatchment"
      }
    ],
    "text": "Outreach Service"
  },
  "subject": {
    "reference": "Patient/61f440c5-fcc6-480f-bfe0-e49bfab71f08"
  },
  "period": {
       "start": "2023-07-05T12:17:56+03:00",
      "end": "2023-07-05T12:17:56+03:00"
  },
  "author": {
    "reference": "Practitioner/6e744577-cf1c-4876-a122-2d0e9c5e0f23"
  },
  "encounter": {
    "reference": "Encounter/example"
  }
}
````

**Comment by samkanga** [comment]
@f-odhiambo the feature required for ZEIR is the one named here as `Out or Area` and which we previously called Out of Catchment Vaccination. See the ticket [here](https://github.com/onaio/fhir-resources/issues/1415). The option for moving patient temporarily and permanently from their cathment area is not in scope for ZEIR. This explain why there was confusion when we were discussion the operation of the Out of Catchment vaccination. What is scoped in this ticket will likely take a longer time to implement and could be parked in favor of the out of catchment vaccination.

**Comment by f-odhiambo** [comment]
There are 2 blockers to this that need discussions 

1. Perform Remote Search as an entry point pre-requisite - https://github.com/opensrp/fhircore/discussions/2163
2. Ability to create/update resources on the server side to perform desired updates 


CC @samkanga @pld @ageryck @dubdabasoduba @rowo

**Comment by dubdabasoduba** [comment]
@f-odhiambo Can we add more information on the **Move Flags**? e.g. the Location ID they are moving from?  This makes it easier to track back to where they came from.

**Reply by f-odhiambo** [reply]
We can do a link in the flag to the encounter and then add the location reference in the Encounter resource, updated the samples for reference

**Reply by dubdabasoduba** [reply]
@f-odhiambo thanks for this do you mind elaborating on how we differentiate between the old Location and the new one? 

NVM. I reviewed the `Encounter.Location` and it kinda makes sense since we can have many locations tagged. We just need to add a code to represent the old location.
